openmdao/components/tests/test_exec_comp.py

```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
         self.assertTrue("balance" in text)
         self.assertTrue("x_1" in text)
 
+
+class TestFunctionRegistration(unittest.TestCase):
+
+    # These 2 tests don't run normally unless you run testflo with a -m "featuretest_*"
+    # I don't want to include them in our normal tests because they don't contain
+    # the _temporary_expr_dict contextmanager (to avoid user confusion)
+    # which keeps the contents of the ExecComp _expr_dict clean
+    def featuretest_register_simple(self):
+        import openmdao.api as om
+
+        om.ExecComp.register("myfunc", lambda x: x * x, complex_safe=True)
+        p = om.Problem()
+        comp = p.model.add_subsystem("comp", om.ExecComp("y = 2 * myfunc(x)"))
+
+        p.setup()
+        p.run_model()
+        J = p.compute_totals(of=['comp.y'], wrt=['comp.x'])
+        assert_near_equal(J['comp.y', 'comp.x'][0][0], 4., 1e-10)
+
+    def featuretest_register_simple_unsafe(self):
+        import openmdao.api as om
+
+        # the following function isn't really complex unsafe, but we'll call it unsafe anyway
+        # for demonstration purposes
+        om.ExecComp.register("unsafe", lambda x: x * x, complex_safe=False)
+        p = om.Problem()
+        comp = p.model.add_subsystem("comp", om.ExecComp("y = 2 * unsafe(x)"))
+
+        # because our function is complex unsafe, we must declare that the partials
+        # with respect to 'x' use 'fd' instead of 'cs'
+        comp.declare_partials('*', 'x', method='fd')
+
+        p.setup()
+        p.run_model()
+        J = p.compute_totals(of=['comp.y'], wrt=['comp.x'])
+        assert_near_equal(J['comp.y', 'comp.x'][0][0], 4., 1e-5)
+
+    def test_register_simple(self):
```

Comment:
I think one test that is missing is one where you have two functions in a list, but only one of them uses a registered function, and we declare fd for that one (maybe it is not complex safe).  The other function should (I think) be computed with the internal cs, so the test should have a tighter tolerance for that one.

Reply:
using manual declare_partials is an all or nothing thing.  Can't mix internal cs with openmdao cs.